vllm/model_executor/layers/fused_moe/fused_moe.py

```diff
@@ -1235,6 +1248,15 @@ def fused_experts_impl(
     if use_int4_w4a16:
         assert hidden_states.shape[1] // 2 == w1.shape[
             2], "Hidden size mismatch"
+    elif use_mxfp4_w4a4:
+        if current_platform.supports_mx() or envs.VLLM_QUARK_EMU_MEM_OPT:
+            # 16bit activation and fp4x2 packed weight
+            assert hidden_states.shape[1] // 2 == w1.shape[
+                2], "hidden size mismatch"
+        else:
+            # 16bit activation and dequantized 16bit weight
+            assert hidden_states.shape[1] == w1.shape[
+                2], "hidden size mismatch"
```

Comment:
I know we discussed this in the dense MXFP PR but I think we should eventually remove the `not envs.VLLM_QUARK_EMU_MEM_OPT` case. Of course it will be slower for inference, but the memory benefit of the lower precision is the expected UX of this format and this complexity is not nice to keep propagating.